diffusion/models/stable_diffusion.py

```diff
@@ -140,13 +149,25 @@ def __init__(self,
 
         # Optional rng generator
         self.rng_generator: Optional[torch.Generator] = None
+        if self.quasirandomness:
+            self.sobol_engine = torch.quasirandom.SobolEngine(dimension=1, scramble=True, seed=self.train_seed)
 
     def _apply(self, fn):
         super(StableDiffusion, self)._apply(fn)
         self.latent_mean = fn(self.latent_mean)
         self.latent_std = fn(self.latent_std)
         return self
 
+    def _generate_quasirandom_timesteps(self, latents: torch.Tensor):
+        # Generate a quasirandom sequence of timesteps equal to the global batch size
+        global_batch_size = latents.shape[0] * dist.get_world_size()
+        timesteps = (len(self.noise_scheduler) * self.sobol_engine.draw(global_batch_size)).squeeze()
```

Comment:
If this was 
```suggestion
        timesteps = ((len(self.noise_scheduler) - 1) * self.sobol_engine.draw(global_batch_size)).squeeze()
```

Would there be a need for the clamp below? To me, the clamp shouldn't be required so having it may mask issues with the code. I think I'd rather an assert if anything

Reply:
See response to the below comment for why it's `len(self.noise_scheduler)` instead of `len(self.noise_scheduler) - 1`

The clamp is because the sobol engine has a small bug where the output is occasionally `1.0` when it is supposed to be [0, 1.0). I don't think it has ever been fixed, but there is a years old bug report [here](https://github.com/pytorch/pytorch/issues/32047). In my testing this happens once every few million timestep generations, when it should happen with probability 0.

You're right that this might mask other issues though. In particular [this one](https://github.com/pytorch/pytorch/issues/33388) which would appear if we generate more than`2^30` timesteps this way. The clamp would still be necessary but I should probably add a check for this and reset the engine if this occurs.

Reply:
Thanks for the explanation!